modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-web/src/main/resources/META-INF/resources/js/ImagePins.js

```diff
@@ -31,72 +33,86 @@ const PIN_ATTRIBUTES = [
 ];
 
 const ImagePins = ({
-	addNewPinState,
-	addPinHandler,
-	cPins,
 	changedScale,
 	enablePanZoom,
 	execZoomIn,
-	handleAddPin,
 	imageSettings,
 	imageURL,
-	namespace,
 	navigationController,
-	removePinHandler,
-	resetZoom,
-	selectedOption,
-	setAddPinHandler,
 	setChangedScale,
-	setCpins,
-	setRemovePinHandler,
-	setResetZoom,
-	setSelectedOption,
-	setShowTooltip,
-	setZoomInHandler,
-	setZoomOutHandler,
 	zoomController,
-	zoomInHandler,
-	zoomOutHandler,
 }) => {
 	const handlers = useRef();
 	const containerRef = useRef();
-	const panZoomRef = useRef();
-
 	const svgRef = useRef(null);
+	const transform = useRef({k: 1, x: 0, y: 0});
+
+	const {
+		addNewPinState,
+		addPinHandler,
+		cPins,
+		isAdmin,
+		removePinHandler,
+		resetZoom,
+		selectedOption,
+		setAddPinHandler,
```

Comment:
Hey @garofalog, What about using [`useReducer` for creating a local store and then just get `dispatch` function](https://pt-br.reactjs.org/docs/hooks-reference.html#usereducer) in DiagramCommonContext?

There is an issue when passing state setters over React context, see:

From https://reactjs.org/docs/hooks-reference.html#usereducer:

> useReducer is usually preferable to useState when you have complex state logic that involves multiple sub-values or when the next state depends on the previous one. useReducer also lets you optimize performance for components that trigger deep updates because you can pass dispatch down instead of callbacks.

Reply:
Hello @diegonvs, and thank you. i was using react context not trying to replace reducers with it, it was just a try but i didn't want to solve anything but only clean up the code.
You're right, using useReducer should resolve a bit this handler's mess (i started doing it simple) so i used that boolean states to call events outside ImagePins component. Also if it works fine i was trying yet to replace in a different way you said that behaviour. Using reducers seems fine to me also because we are facing different topics about others related components and this kind of refactor could help us even because we should update a liferay state outiside react components to make our three react-component comunicate in LR.

If something it's not clear please ask me, we decided this behaviour more or less, today.
Going to the end we want to refactor in this way but we can't do it in this pr, we'll create a subtask to work on it later.
So i'll remove the useContext if seems to be not useful.